encoding/cbor/decode_test.go

```diff
@@ -0,0 +1,1334 @@
+		},
+		"{_ } / no break": {
+			[]byte{5<<5 | 31},
+			"expected break marker",
+		},
+		"{_ } / non-string key": {
+			[]byte{5<<5 | 31, 0},
+			"unexpected major type 0 for map key",
+		},
+		"{_ } / invalid key": {
+			[]byte{5<<5 | 31, 3<<5 | 24, 1},
+			"slice len 1 greater than remaining buf len",
+		},
+		"{_ } / invalid value": {
+			[]byte{5<<5 | 31, 3<<5 | 3, 0x66, 0x6f, 0x6f, 0<<5 | 24},
+			"arg len 1 greater than remaining buf len",
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			_, _, err := decode(c.In)
+			if err == nil {
+				t.Errorf("expect err %s", c.Err)
+			}
+			if aerr := err.Error(); !strings.Contains(aerr, c.Err) {
+				t.Errorf("expect err %s, got %s", c.Err, aerr)
+			}
+		})
+	}
+}
+
+func TestDecode_InvalidTag(t *testing.T) {
+	for name, c := range map[string]struct {
+		In  []byte
+		Err string
+	}{
+		"invalid value": {
+			[]byte{6<<5 | 1, 0<<5 | 24},
+			"arg len 1 greater than remaining buf len",
+		},
+		"eof": {
+			[]byte{6<<5 | 1},
+			"unexpected end of payload",
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			_, _, err := decode(c.In)
+			if err == nil {
+				t.Errorf("expect err %s", c.Err)
+			}
+			if aerr := err.Error(); !strings.Contains(aerr, c.Err) {
+				t.Errorf("expect err %s, got %s", c.Err, aerr)
+			}
+		})
+	}
+}
+
+func TestDecode_Atomic(t *testing.T) {
+	for name, c := range map[string]struct {
+		In     []byte
+		Expect Value
+	}{
+		"uint/0/min": {
+			[]byte{0<<5 | 0},
+			Uint(0),
+		},
+		"uint/0/max": {
+			[]byte{0<<5 | 23},
+			Uint(23),
+		},
+		"uint/1/min": {
+			[]byte{0<<5 | 24, 0},
+			Uint(0),
+		},
+		"uint/1/max": {
+			[]byte{0<<5 | 24, 0xff},
+			Uint(0xff),
+		},
+		"uint/2/min": {
+			[]byte{0<<5 | 25, 0, 0},
+			Uint(0),
+		},
+		"uint/2/max": {
+			[]byte{0<<5 | 25, 0xff, 0xff},
+			Uint(0xffff),
+		},
+		"uint/4/min": {
+			[]byte{0<<5 | 26, 0, 0, 0, 0},
+			Uint(0),
+		},
+		"uint/4/max": {
+			[]byte{0<<5 | 26, 0xff, 0xff, 0xff, 0xff},
+			Uint(0xffffffff),
+		},
+		"uint/8/min": {
+			[]byte{0<<5 | 27, 0, 0, 0, 0, 0, 0, 0, 0},
+			Uint(0),
+		},
+		"uint/8/max": {
+			[]byte{0<<5 | 27, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
+			Uint(0xffffffff_ffffffff),
+		},
+		"negint/0/min": {
+			[]byte{1<<5 | 0},
+			NegInt(1),
+		},
+		"negint/0/max": {
+			[]byte{1<<5 | 23},
+			NegInt(24),
+		},
+		"negint/1/min": {
+			[]byte{1<<5 | 24, 0},
+			NegInt(1),
+		},
+		"negint/1/max": {
+			[]byte{1<<5 | 24, 0xff},
+			NegInt(0x100),
+		},
+		"negint/2/min": {
+			[]byte{1<<5 | 25, 0, 0},
+			NegInt(1),
+		},
+		"negint/2/max": {
+			[]byte{1<<5 | 25, 0xff, 0xff},
+			NegInt(0x10000),
+		},
+		"negint/4/min": {
+			[]byte{1<<5 | 26, 0, 0, 0, 0},
+			NegInt(1),
+		},
+		"negint/4/max": {
+			[]byte{1<<5 | 26, 0xff, 0xff, 0xff, 0xff},
+			NegInt(0x100000000),
+		},
+		"negint/8/min": {
+			[]byte{1<<5 | 27, 0, 0, 0, 0, 0, 0, 0, 0},
+			NegInt(1),
+		},
+		"negint/8/max": {
+			[]byte{1<<5 | 27, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe},
```

Comment:
So essentially I chose to build in the -1 bias from type 1 into the codec layer. Since the range of type1 is [-2^64,-1], the wrap to zero that comes from a type1 of 64 1s would basically be that implicit 65-bit left bound.

In practice remember the caller doesn't ever see it, it's something the generated deserializer would reject since -2nd ^64 doesn't fit in long.

We could drop the implicit bias here and instead have the generated deserializers reintroduce it but I think we'd be trading up in terms of code size (it's a special case on `NegInt(0)` vs a bunch of +1s).

Reply:
Maybe to clarify, this code exists in the in-progress deserializer, this is basically what happens when we're deserializing to something that's modeled as a `long`.

```go
// AsInt64 coerces a Value to its int64 representation if possible.
func AsInt64(v Value) (int64, error) {
    const max64 = 0x7fffffff_ffffffff

    switch vv := v.(type) {
    case Uint:
        if vv > max64 {
            return 0, fmt.Errorf("cbor uint %d exceeds max int64 value", vv) 
        }   
        return int64(vv), nil 
    case NegInt:
        if vv > max64+1 || vv == 0 { // -1 bias means NegInt(0) is actually -2^64
            return 0, fmt.Errorf("cbor negint %d exceeds min int64 value", vv) 
        }   
        return -int64(vv), nil 
    }   
    return 0, fmt.Errorf("unexpected value type %T", v)
}
```